src/common/colorspaces_inline_conversions.h

```diff
@@ -271,20 +271,40 @@ static inline void dt_Lab_to_linearRGB(
     RGB[r] = cmatrix_row0[r] * XYZ[0] + cmatrix_row1[r] * XYZ[1] + cmatrix_row2[r] * XYZ[2];
 }
 
+/* We ensure safe XYZ data first.
+   The calculation for black would fail with NaNs as result.
+   According to http://www.brucelindbloom.com/index.html?Eqn_XYZ_to_xyY.html
+   we would want the pipes chromaticity coordinates of the reference white.
+   To avoid passing the specific whitepoint coordinates we define two variants for D50 and D65
+*/
+
 #ifdef _OPENMP
-#pragma omp declare simd aligned(xyY, XYZ:16)
+#pragma omp declare simd aligned(xyY, sXYZ:16)
 #endif
-static inline void dt_XYZ_to_xyY(const dt_aligned_pixel_t XYZ, dt_aligned_pixel_t xyY)
+static inline void dt_D65_XYZ_to_xyY(const dt_aligned_pixel_t sXYZ, dt_aligned_pixel_t xyY)
 {
-  const gboolean black = XYZ[0] == 0.0f && XYZ[1] == 0.0f && XYZ[2] == 0.0f;
-  /* the calculation for black would fail with NaNs as result.
-     According to http://www.brucelindbloom.com/index.html?Eqn_XYZ_to_xyY.html
-     we would want the pipes chromaticity coordinates of the reference white.
-     The best guess as a fallback is from D65_xyY so
-  */
+  static const dt_aligned_pixel_t zero = { 0.0f, 0.0f, 0.0f, 0.0f };
+  dt_aligned_pixel_t XYZ;
+  dt_vector_max(XYZ, sXYZ, zero);
```

Comment:
In general these changes seem good. However this clip is something that I'm slightly concerned about.

There are some RGB color spaces that have negative values for some of the primaries. Such as ACES AP0 or some camera sensor RGB spaces.

In the call sites changed in this PR, it seems we would likely not be using one of those color spaces, but something to keep in mind anyway. I would personally prefer not clipping here, in favour of the least surprise.

Since this seemed to relate to fixing an issue in color eq when fed with -0.0 - is there some issue to fix upstream? What modules were in the pipe before color eq?

Reply:
Perfect to see you commenting here! I don't remember exactly all color maths fixes done over the last weeks, all were done while investigating colorbalancergb and the color equalizer working in UCS mode. 
1. div by zero
2. div by close-to-zero resulting in +inf -inf results or at least in values exceeding the defined range for UCS
3. div by +inf -inf (or very close to) resulting in possibly extreme small negative numbers
4. powf(a, b) usage with a < 0 and various b

In UCS mode we could have neg L_Star values or so high resulting in wrong back transforms.

There is one example you might be interested in https://github.com/darktable-org/darktable/pull/16071 the single-pixel bright artifacts. We do an initial clipneg (L727) but when doing the XYZ_to_xyY conversion we have small negs as XYZ.

As we use predefined matrices using const single float data i a) am not fully sure about their 'correctness' b) don't know if multiple ops following each other and maybe compiler flags might lead to extremely-small-negs. At least on intel/amd OpenCL we have had a number of issues from such conditions.

Before using the public `cmsCIExyY D65xyY` data all-over-the-place we had slightly different data in use, one more reason to doubt the given consts in all matrix ops are "absolutely correct"

> I would personally prefer not clipping here, in favour of the least surprise.
Then we would have to "manually clip" before XYZ_to_xyY conversions, missing that resulted in artifacts in colorbalancergb for some dt versions ... don't know